src/backend/distributed/planner/multi_planner.c

```diff
@@ -117,38 +187,154 @@ multi_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
+{
+	MultiPlan *physicalPlan = NULL;
+	PlannedStmt *resultPlan = NULL;
+	DeferredErrorMessage *error = NULL;
+	bool hasUnresolved = false;
+
+	if (HasUnresolvedExternParamsWalker((Node *) query, boundParams))
+	{
+		hasUnresolved = true;
+	}
+
+	if (IsModifyCommand(query))
+	{
+		/*
+		 * Modifications are always routed through the same executor. As
+		 * there's currently no other way to plan these, error out if the
+		 * query is unsupported.
+		 */
+		error = CreateModifyPlan(&physicalPlan, originalQuery, query,
+								 restrictionContext);
+
+		/*
+		 * Error out if there's an error, unless the error was possibly
+		 * triggered by missing parameters.  In that case an alternative plan
+		 * will be generated with bound parameters (see below).
+		 */
+		if (error && !hasUnresolved)
+		{
+			RaiseDeferredError(error, ERROR);
+		}
+	}
+	else
+	{
+		/*
+		 * For select query we, if router executor is enabled, first try to
+		 * plan the query as a router query. If not supported, otherwise try
+		 * the full blown plan/optimize/physical plan process needed to
+		 * produce distributed query plans.
+		 */
+		if (EnableRouterExecution)
+		{
+			error = CreateRouterPlan(&physicalPlan, originalQuery, query,
+									 restrictionContext);
+
+			if (error)
+			{
+				RaiseDeferredError(error, DEBUG1);
+			}
+		}
+
+		/*
+		 * Router didn't yield plan, try the full distributed planner. As
+		 * real-time/task-tracker don't support prepared statement parameters,
+		 * skip planning in that case (we'll later trigger error in that
+		 * case).
+		 */
+		if (!physicalPlan && !hasUnresolved)
+		{
+			/* Create and optimize logical plan */
+			MultiTreeRoot *logicalPlan = MultiLogicalPlanCreate(query);
+			MultiLogicalPlanOptimize(logicalPlan);
+
+			/*
+			 * This check is here to make it likely that all node types used in
+			 * Citus are dumpable. Explain can dump logical and physical plans
+			 * using the extended outfuncs infrastructure, but it's infeasible to
+			 * test most plans. MultiQueryContainerNode always serializes the
+			 * physical plan, so there's no need to check that separately.
+			 */
+			CheckNodeIsDumpable((Node *) logicalPlan);
+
+			/* Create the physical plan */
+			physicalPlan = MultiPhysicalPlanCreate(logicalPlan);
+		}
 
 		/*
-		 * This check is here to make it likely that all node types used in
-		 * Citus are dumpable. Explain can dump logical and physical plans
-		 * using the extended outfuncs infrastructure, but it's infeasible to
-		 * test most plans. MultiQueryContainerNode always serializes the
-		 * physical plan, so there's no need to check that separately.
+		 * Error out if none of the planners resulted in a usable plan error,
+		 * unless the error was possibly triggered by missing parameters.  In
+		 * that case an alternative plan will be generated with bound
+		 * parameters (see below).
 		 */
-		CheckNodeIsDumpable((Node *) logicalPlan);
+		if (!physicalPlan && error && !hasUnresolved)
```

Comment:
I think the `!physicalPlan` is redundant here. Also, same comment as for the modify part.